packages/core/test/smoke-valibot.test.ts

```diff
@@ -629,3 +629,115 @@ describe("extending presets", () => {
     });
   });
 });
+
+describe("createFinalSchema", () => {
```

Comment:
how does this work with `extends`? can you add some tests for that

Reply:
can do - extends just assigns all of the properties to the final result from the schema, like it did before

Reply:
what I was thinking of was more like what if you do some combinator that accesses server stuff, will that cause an "invalid server access" or not? I'd think not since you're not giving the proxy as the argument there right? also what are the types in that callback like, server variables will be undefined when the combinator runs on the client etc etc

Reply:
had a brainwave and opened #316 to fix #240

coming back to this - schemas have always been allowed to access any key they want, though Typescript-wise that'll only include the ones that have already been checked with the current schema (not any extended ones).
![image](https://github.com/user-attachments/assets/0690cde1-2d05-43c1-bd0f-0a0d2fc369c5)

Reply:
> schemas have always been allowed to access any key they want

No since accessing a server var on the client would throw an error - so e.g if you'd chain on a `toLower()` you'd get the invalid access before. It seems here we'd get a "Cannot access property toLower of undefined" which feels suboptimal - but maybe I'm overthinking it

Reply:
schemas have never been run on the output of presets, they were always run before it. they have always been passed the raw runtimeEnv option, and any presets were merged into the parsed result.

Reply:
I understand the concern, but I don't think there's a good way to expose type-wise that the server variables might not be defined to the `createFinalSchema` callback. Best case, the schema result *also* reflects the server variables as optional (whereas they're currently always defined), worst case the inference breaks completely. We also can't wrap the intermediate parsed value in a Proxy since we don't have access to it.

We could *possibly* pass some sort of `makeSafe` function to the user that would make a Proxy, but that would need to be opt in.
```ts
createObjectSchema: (shape, { isServer, makeSafe }) => z.object(shape).refine((_env) => {
  const env = makeSafe(_env) // wraps in Proxy
  if (!isServer) env.SERVER_ENV // throws
  // ...
})
```